funtools/man/man1/funhead.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.32
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "funhead 1"
.TH funhead 1 "April 14, 2011" "version 1.4.5" "SAORD Documentation"
.SH "NAME"
funhead \- display a header in a Funtools file
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBfunhead\fR  [\-a] [\-s] [\-t] [\-L] <iname> [oname ename]
.SH "OPTIONS"
.IX Header "OPTIONS"
.Vb 4
\&  \-a    # display all extension headers
\&  \-s    # display 79 chars instead of 80 before the new-line
\&  \-t    # prepend data type char to each line of output
\&  \-L    # output in rdb/starbase list format
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBfunhead\fR displays the \s-1FITS\s0 header parameters in the specified 
\&\s-1FITS\s0 Extension.
.PP
The first argument to the program specifies the Funtools input file
to display.  If \*(L"stdin\*(R" is specified, data are read from
the standard input. Funtools Bracket
Notation is used to specify particular \s-1FITS\s0 extension to process.
Normally, the full 80 characters of each header card is output,
followed by a new\-line. 
.PP
If the \fB\-a\fR switch is specified, the header from each \s-1FITS\s0
extensions in the file is displayed. Note, however, that the \fB\-a\fR
switch does not work with \s-1FITS\s0 files input via stdin. We hope to
remove this restriction in a future release.
.PP
If the \fB\-s\fR switch is specified, only 79 characters are output
before the new\-line. This helps the display on 80 character terminals.
.PP
If the \fB\-t\fR switch is specified, the data type of the parameter
is output as a one character prefix, followed by 77 characters of the
param.  The parameter data types are defined as: \s-1FUN_PAR_UNKNOWN\s0
('u'), \s-1FUN_PAR_COMMENT\s0 ('c'), \s-1FUN_PAR_LOGICAL\s0 ('l'), \s-1FUN_PAR_INTEGER\s0
('i'), \s-1FUN_PAR_STRING\s0 ('s'), \s-1FUN_PAR_REAL\s0 ('r'), \s-1FUN_PAR_COMPLEX\s0 ('x').
.PP
If the \fB\-L\fR (rdb table) switch is used, the output will conform
to starbase/rdb data base list format.
.PP
For example to display the \s-1EVENTS\s0 extension (binary table):
.PP
.Vb 17
\&  [sh] funhead "foo.fits[EVENTS]"
\&  XTENSION= 'BINTABLE'            /  FITS 3D BINARY TABLE                      
\&  BITPIX  =                    8  /  Binary data                               
\&  NAXIS   =                    2  /  Table is a matrix                         
\&  NAXIS1  =                   20  /  Width of table in bytes                   
\&  NAXIS2  =                30760  /  Number of entries in table                
\&  PCOUNT  =                    0  /  Random parameter count                    
\&  GCOUNT  =                    1  /  Group count                               
\&  TFIELDS =                    7  /  Number of fields in each row              
\&  EXTNAME = 'EVENTS  '            /  Table name                                
\&  EXTVER  =                    1  /  Version number of table                   
\&  TFORM1  = '1I      '            /  Data type for field                       
\&  TTYPE1  = 'X       '            /  Label for field                           
\&  TUNIT1  = '        '            /  Physical units for field                  
\&  TFORM2  = '1I      '            /  Data type for field                       
\&    etc. ...
\&  END
.Ve
.PP
To display the third header:
.PP
.Vb 14
\&  [sh] funhead "foo.fits[3]"
\&  XTENSION= 'BINTABLE'            /  FITS 3D BINARY TABLE                      
\&  BITPIX  =                    8  /  Binary data                               
\&  NAXIS   =                    2  /  Table is a matrix                         
\&  NAXIS1  =                   32  /  Width of table in bytes                   
\&  NAXIS2  =                   40  /  Number of entries in table                
\&  PCOUNT  =                    0  /  Random parameter count                    
\&  GCOUNT  =                    1  /  Group count                               
\&  TFIELDS =                    7  /  Number of fields in each row              
\&  EXTNAME = 'TGR     '            /  Table name                                
\&  EXTVER  =                    1  /  Version number of table                   
\&  TFORM1  = '1D      '            /  Data type for field                       
\&    etc. ...
\&  END
.Ve
.PP
To display the primary header (i.e., extension 0):
.PP
.Vb 8
\&  sh> funhead "coma.fits[0]"
\&  SIMPLE  =                    T /STANDARD FITS FORMAT                         
\&  BITPIX  =                   16 /2-BYTE TWOS-COMPL INTEGER                    
\&  NAXIS   =                    2 /NUMBER OF AXES                               
\&  NAXIS1  =                  800 /                                             
\&  NAXIS2  =                  800 /                                             
\&  DATATYPE= 'INTEGER*2'          /SHORT INTEGER                                
\&  END
.Ve
.PP
The funhead program also can edit (i.e. add, delete, or modify) or
display individual headers parameters. Edit mode is signalled by the
presence of two additional command-line arguments: output file and
edit command file, in that order. Edit mode acts as a filter: the
output file will contain the entire input \s-1FITS\s0 file, including other
extensions. The edit command file can be \*(L"stdin\*(R", in which case edit
command are read from the standard input.
.PP
The edit command file contains parameter comments (having '#' in the
first column) and delete and assignment(modify or add) operations.  A
delete operation is specified by preceding the parameter name with a
minus sign \*(L"\-\*(R".  A display operation (very useful in interactive
sessions, i.e., where the edit commands are taken from stdin) is
specified by preceding the parameter name with a question mark \*(L"?\*(R". In
either case, a parameter value need not be specified.  An assignment
operation is specified in the same two ways that a parameter is
specified in a text header (but without the comment character that
precedes header params), i.e.:
.IP "\(bu" 4
FITS-style comments have an equal sign \*(L"=\*(R" between the keyword and
value and an optional slash \*(L"/\*(R" to signify a comment. The strict \s-1FITS\s0
rules on column positions are not enforced.
.IP "\(bu" 4
Free-form comments can have an optional colon separator between the
keyword and value. In the absence of quote, all tokens after the
keyword are part of the value, i.e. no comment is allowed.
.PP
For example, the following interactive session checks for the
existence of parameters, adds new parameters, modifies them, and
modifies and deletes existing parameters:
.PP
.Vb 20
\&  sh$ ./funhead snr.ev foo.fits -
\&  # look for FOO1
\&  ? FOO1
\&  WARNING: FOO1 not found
\&  # add new foo1
\&  FOO1 = 100
\&  # add foo2
\&  FOO2 = 200
\&  # reset foo1 to a different value
\&  FOO1 \-1
\&  # delete foo2
\&  \-FOO2
\&  # change existing value
\&  EXTVER 2
\&  ? XS-SORT
\&  XS-SORT = 'EOF     '            /  type of event sort
\&  # delete existing value
\&  \-XS-SORT
\&  # exit
\&  ^D
.Ve
.PP
See Column-based Text Files
for more information about header parameter format.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
See funtools(7) for a list of Funtools help pages


ChangeLog
2016-01-12 waj

	* upgraded TEA to 3.9
	* numerous compiler fixes for 64bit

2008-11-14  Jeff Hobbs  <jeffh@ActiveState.com>

	**** 2.10 TAGGED FOR RELEASE ****

	* generic/tkTable.c: Set NO_XSETCLIP universally as Tk 8.5/Xft
	seems to ignore XSetClipRectangles.  This may slow things down a
	little, but draws correctly. [Bug 1805350]

	* UPGRADING.txt (removed): no longer need this old v1 upgrade doc.

	* README.txt: remove refs to mac classic stuff, update notes

	* configure, configure.in, tclconfig/tcl.m4: Update to TEA 3.7
	* tests/tkTable.test, unix/tktable.spec, generic/version.h:
	* demos/loadtable.tcl: Update version to 2.10
	* Makefile.in: update dist target, include library/tktable.py

	* demos/tktable.py: note that this version is outdated
	* library/tktable.py (NEW): Added new version of tktable wrapper
	for Python/Tkinter from GPolo.  Not 100% compatible with previous
	version. [Patch 2244167]

	* generic/tkTableTag.c (Table_TagCmd): remove interp->result set
	work-around for old bug as i->result is no longer exposed in 8.6.

	* generic/tkTable.c:     Use safe Tcl_SetObjResult instead of trying
	* generic/tkTableCell.c: to set existing objResult value.
	* generic/tkTableCmds.c: 

2006-07-10  Jeff Hobbs  <jeffh@ActiveState.com>

	* license.txt (bourbon_ware): Be less restrictive in the
	bourbon_ware clause.

2006-01-25  Jeff Hobbs  <jeffh@ActiveState.com>

	* configure.in, configure, tclconfig/tcl.m4: update to TEA 3.5

2006-01-22  Jeff Hobbs  <jeffh@ActiveState.com>

	* tclconfig/tcl.m4, configure: update to TEA 3.4 interim

2005-12-13  Jeff Hobbs  <jeffh@ActiveState.com>

	* generic/tkTableCmds.c (Table_ActivateCmd): remove potential
	undefined behavior warning.

2005-12-02  Jeff Hobbs  <jeffh@ActiveState.com>

	* tclconfig/tcl.m4, configure.in, configure: update to TEA 3.4

2005-05-11  Jeff Hobbs  <jeffh@ActiveState.com>

	* Makefile.in (html): switch to groff to fix botched html output.
	* doc/tkTable.html: regen [Bug 1198478]
	* doc/tkTable.n: use '' around \\ items.

2005-03-18  Jeff Hobbs  <jeffh@ActiveState.com>

	* Makefile.in (AR): use @AR@, improve html doc target
	* configure, configure.in, tclconfig/tcl.m4: TEA 3.2 patch update

	* doc/tkTable.n, doc/kTable.html: doc cleanup

2005-02-01  Jeff Hobbs  <jeffh@ActiveState.com>

	* configure, configure.in, tclconfig/tcl.m4: update to TEA 3.2

2004-11-18  Jeff Hobbs  <jeffh@ActiveState.com>

	* library/tkTable.tcl (::tk::table::SelectAll): remove invalid ref
	to HandleType proc. [Bug 1068874] (ade)

2004-07-22  Jeff Hobbs  <jeffh@ActiveState.com>

	**** 2.9 TAGGED FOR RELEASE ****

	* configure, tclconfig/tcl.m4: update tcl.m4 TEA_PATH_X that adds
	Aqua build recognition and TEA_WINDOWINGSYSTEM variable.

2004-07-20  Jeff Hobbs  <jeffh@ActiveState.com>

	* generic/version.h, unix/tktable.spec:    updated for v2.9
	* demos/loadtable.tcl, tests/tkTable.test:

	* Makefile.in, configure.in, configure, tclconfig/tcl.m4: update
	to TEA 3.1, use CFLAGS instead of EXTRA_CFLAGS and bump the
	version number of Tktable to 2.9.

	* generic/tkTable.h: add #include <ctype.h>

	* generic/tkTableCell.c (TableGetCellValue, TableSetCellValue): 
	Do faster cache checks when DATA_CACHE is the only data source.
	In fallback from command that errors to array var, retrieve array
	value immediately on current get|set request.
	When setting the value with cache and array sources, don't do the
	caching branch as the array trace will do that already.
	(TableMoveCellValue): Do the efficient cache value management when
	cache is the only data source, otherwise the regular fallback has
	all the necessary checks. [Bug #919872]

	* generic/tkTable.c (TableVarProc): allow setting NULL value in
	cache hash from array trace.

2004-06-10  Jeff Hobbs  <jeffh@ActiveState.com>

	* generic/tkTableWin.c (EmbWinDisplay): make min window displayed
	have width/height 1 pixel min (was 3px)

	* generic/tkTable.c (TableDisplay): Don't draw when we don't have
	the space for it. [Bug 747038]

	* doc/tkTable.html, doc/tkTable.n: improve docs to note
	requirement of specifying a data source. [Bug 963204]

2004-02-17  Jeff Hobbs  <jeffh@ActiveState.com>

	* generic/tkTableCmds.c (Table_ActivateCmd): test implementation
	of an "unactivate" ($table activate "") command to remove the
	active cell. (becroft)

2004-02-13  Jeff Hobbs  <jeffh@ActiveState.com>

	* library/tkTable.tcl (::tk::table::Button1): account for what
	resizeborders are set when selecting cell [Bug 876320] (ferenc) 
	(<Prior> <Next>): activate topleft, not @0,0 (accounts for title
	rows) [Bug 879347] (ferenc)

2003-12-10  Jeff Hobbs  <jeffh@ActiveState.com>

	* Makefile.in: updated to TEA3 spec
	* configure:
	* configure.in:
	* generic/tkTable.c:
	* generic/tkTableInitScript.h:
	* generic/version.h:
	* mac/mac_tkTable.r:
	* tclconfig/tcl.m4:

	* win/makefile.vc: minor updates, probably still out of date

	* tclconfig/ChangeLog (removed): not necessary

	* generic/tkTable.c: added Tk_ClassProcs and TableWorldChanged
	proc if built against 8.4

	* generic/tkTable.h: add 8.4 ifdef, remove ANSI_ARGS usage

2003-07-16  Jeff Hobbs  <jeffh@ActiveState.com>

	* doc/tkTable.n: add note about -window "" behavior

2003-04-10  Jeff Hobbs  <jeffh@ActiveState.com>

	* configure: regen
	* tclconfig/tcl.m4: updated

	* library/tkTable.tcl (BeginExtend): avoid error in tables that
	have no anchor index yet. (renshaw)

2003-04-09  Jeff Hobbs  <jeffh@ActiveState.com>

	* configure: regen
	* tclconfig/tcl.m4: Updated to newest tcl.m4, (WinCE support)

	* demos/loadtable.tcl: redefine 'puts' on Windows CE

	* generic/tkTable.c (TableDisplay): correctly copy clipwindow area
	for NO_XSETCLIP case and use that when UNDER_CE as well.  Also add
	support for drawing ellipsis in NO_XSETCLIP case.
	Comment out OffsetClipRgn call - it was a noop.

2003-04-04  Andreas Kupries  <andreask@activestate.com>

	* configure.in:
	* tclconfig/tcl.m4: Updated to newest tcl.m4, regenerated
	  configure's.


funtools/funhead.c
/*
 *	Copyright (c) 1999-2005 Smithsonian Astrophysical Observatory
 */

#include <funtoolsP.h>

#define MAXROW 8192
static int maxrow=MAXROW;

#define BLANK "                                                                                "

extern char *optarg;
extern int optind;

#ifdef ANSI_FUNC
static int 
_FunHeadEdit(char *iname, char *oname, char *ename)
#else
  static int _FunHeadEdit(iname, oname, ename)
     char *iname;
     char *oname;
     char *ename;
#endif
{
  int i;
  int type;
  int dim2;
  int action=1;
  int got=0;
  int put=0;
  int pgot=0;
  char *ev;
  char *eptr=NULL;
  char ebuf[SZ_LINE];
  char key[SZ_LINE];
  char val[SZ_LINE];
  char com[SZ_LINE];
  char tfile[SZ_LINE];
  char tbuf[SZ_LINE];
  char *buf=NULL;
  Fun ifun=NULL;
  Fun ofun=NULL;
  GIO efd=NULL;
  FITSHead header=NULL, theader=NULL;
  FITSCard card;

  /* sanity check */
  if( !*iname || !*ename ) return 0;

  /* temp file if no output name is given */
  if( !oname ){
    if( !mkrtemp("fun", ".fits", tfile, SZ_LINE, 0) ){
      gerror(stderr, "can't open temp file for param editing\n");
    }
    oname = tfile;
  }

  /* open file(s) */
  if( !(efd = gopen(ename, "r")) ){
    gerror(stderr, "can't open edit file: %s\n", ename);
  }
  if( !(ifun = FunOpen(iname, "rC", NULL)) ){
    gerror(stderr, "can't FunOpen input file (or find extension): %s\n",
	   iname);
  }
  if( !(ofun = FunOpen(oname, "w", ifun)) ){
    gerror(stderr, "can't FunOpen output file: %s\n",
	   oname);
  }

  /* we write into the temp space of the output header */
  FunInfoGet(ifun, FUN_HEADER,  &header, FUN_TYPE, &type, 0);
  FunInfoGet(ofun, FUN_THEADER, &theader, 0);

  /* read and process the edit commands */
  while( ggets (efd, ebuf, SZ_LINE) ){
    eptr = ebuf;
    /* skip comments and blank lines */
    if( (*eptr == '#') || (*eptr == '\n') ) continue;
    /* skip whitespace */
    while( isspace((int)*eptr) ) eptr++;
    /* check for action type */
    switch((int)*eptr){
    case '-':
      action = 1;
      eptr++;
      break;
    case '?':
      action = 2;
      eptr++;
      break;
    default:
      action = 3;
      break;
    }
    /* analyze string */
    pgot = FunTextParam(NULL, eptr, key, val, com, SZ_LINE);
    /* key is always upper case */
    cluc(key);
    /* process action for this parameter */
    switch(action){
    case 1:
      /* delete the parameter */
      if( (card=ft_headfind(theader, key, 0, 0)) && *card->c ){
	ft_headdel(theader, key, 0);
      }
      else if( (card=ft_headfind(header, key, 0, 0)) && *card->c  ){
	ft_headdel(header, key, 0);
      }
      else{
	fprintf(stdout, "WARNING: %s not found\n", key);
      }
      break;
    case 2:
      /* display the parameter */
      if( ((card=ft_headfind(theader, key, 0, 0)) && *card->c)  ||
 	  ((card=ft_headfind(header, key, 0, 0)) && *card->c)   ){
	strncpy(tbuf, card->c, FT_CARDLEN);
	tbuf[FT_CARDLEN] = '\0';
	fprintf(stdout, "%s\n", tbuf);
      }
      else{
	fprintf(stdout, "WARNING: %s not found\n", key);
      }
      fflush(stdout);
      break;
    case 3:
      /* add or modify the parameter */
      FunTextParamHeader(theader, eptr, key, val, com, pgot);
      break;
    default:
      break;
    }
  }
  if( efd )  gclose(efd);

  /* copy data */
  switch(type){
  case FUN_IMAGE:
  case FUN_ARRAY:
    FunInfoGet(ifun, FUN_SECT_DIM2, &dim2, 0);
    /* set dim2 to 1 for a 1D image */
    if( dim2 == 0 ) dim2 = 1;
    /* copy the image from input to output */
    for(i=1; i<=dim2; i++){
      /* extract image row by row */
      if( !(buf = FunImageRowGet(ifun, buf, i, i, NULL)) )
	gerror(stderr, "can't FunImageRowGet: %s\n", iname);
      /* write output image row by row */
      if( !FunImageRowPut(ofun, buf, i, i, 0, 0, 0, NULL) )
	gerror(stderr, "can't FunImageRowPut: %s\n", oname);
    }
    if( buf ) xfree(buf);
    break;
  case FUN_TABLE:
  case FUN_EVENTS:
    /* copy the events from input to output */
    while( (ev = FunTableRowGet(ifun, NULL, maxrow, NULL, &got)) ){
      if( (put=FunTableRowPut(ofun, ev, got, 0, NULL)) != got ){
	gerror(stderr, "expected to write %d rows; only wrote %d\n",
	       got, put);
      }
      if( ev ) xfree(ev);
    }
    break;
  }

  /* flush and close files */
  if( ofun ) FunClose(ofun);
  if( ifun ) FunClose(ifun);
  return 0;
}

#ifdef ANSI_FUNC
static void
usage (char *fname)
#else
static void usage(fname)
     char *fname;
#endif
{
  fprintf(stderr,
	  "usage: %s [-a] [-s] [-t] [-L] iname[ext] [oname ename]\n", 
	  fname);
  fprintf(stderr, "optional switches:\n");
  fprintf(stderr, "  -a  # display all extension headers\n");
  fprintf(stderr, "  -s  # display 79 chars instead of 80 before the new-line \n");
  fprintf(stderr, "  -t  # prepend data type char to each line of output\n");
  fprintf(stderr, "  -L  # output in rdb/starbase list format\n");
  fprintf(stderr, "\n(version: %s)\n", FUN_VERSION);
  exit(1);
}

#ifdef ANSI_FUNC
int 
main (int argc, char **argv)
#else
int
main(argc, argv)
     int argc;
     char **argv;
#endif
{
  int c;
  int i;
  int args;
  int got;
  int clen=0;
  int extno=0;
  int doall=0;
  int dotype=0;
  int dolist=0;
  int rcode=0;
  char *s, *t;
  char *ind;
  char *fmt="%.80s\n";
  char *iname=NULL;
  char *oname=NULL;
  char *ename=NULL;
  char *cname=NULL;
  Fun fun;

  /* exit on gio errors */
  if( !getenv("GERROR")  )
    setgerror(2);

  /* we want the args in the same order in which they arrived, and
     gnu getopt sometimes changes things without this */
  putenv("POSIXLY_CORRECT=true");

  /* process switch arguments */
  while ((c = getopt(argc, argv, "astL")) != -1){
    switch(c){
    case 'a':
      doall = 1;
      break;
    case 's':
      fmt = "%.79s\n";
      break;
    case 't':
      dotype = 1;
      fmt = "%.77s\n";
      break;
    case 'L':
      dolist = 1;
      break;
    }
  }

  /* get maxrow,if user-specified */
  if( (s=(char *)getenv("FUN_MAXROW")) != NULL )
    maxrow = atoi(s);

  /* check for required arguments */
  args = argc - optind;
  if( (args != 1) && (args != 3) ) usage(argv[0]);

  /* input file name */
  iname = xstrdup(argv[optind+0]);

  /* edit expression right here and exit */
  if( args >= 2 ){
    if( args >= 3 ){
      oname = xstrdup(argv[optind+1]);
      ename = xstrdup(argv[optind+2]);
    }
    else{
      ename = xstrdup(argv[optind+1]);
    }
    rcode = _FunHeadEdit(iname, oname, ename);
    goto done;
  }
    
  /* if we are doing all extensions, remove bracket specifcation from base */
  if( doall ){
    if( (ind=strchr(iname, '[')) || (ind=strchr(iname, ',')) ){
      *ind = '\0';
    }
  }

  /* set up current name */
  clen = strlen(iname) + SZ_LINE;
  cname = xcalloc(clen, sizeof(char));

  /* set up the first extension */
  if( doall ){
    snprintf(cname, clen, "%s[%d]", iname, extno);
    extno++;
  }
  else{
    strncpy(cname, iname, clen);
  }

again:
  /* open the input FITS extension  */
  if( !(fun = FunOpen(cname, "r", NULL)) ){
    /* if we are doing all extension, an error means we are done */
    if( doall )
      return(0);
    /* not doing all, so if no extension was given, try primary */
    if( !strchr(iname, '[') && !strchr(iname, ',') ){
      snprintf(cname, clen, "%s[0]", iname);
      if( !(fun = FunOpen(cname, "r", NULL)) )
	gerror(stderr, "can't FunOpen input file (or find extension): %s\n",
	       iname);
    }
    else{
      gerror(stderr, "can't FunOpen input file (or find extension): %s\n",
	     iname);
    }
  }


  if ( dolist ) {
      fprintf(stdout, "\n");
      fprintf(stdout, "FITSNAME\t%s\n", iname);
  } else {
      /* get extension name */
      if( (s=FunParamGets(fun, "EXTNAME", 0, NULL, &got)) ){
	  fprintf(stdout, "\t\tExtension: %s", s);
	  if( (t=FunParamGets(fun, "HDUNAME", 0, NULL, &got)) )
	      fprintf(stdout, "/%s", t);
	  fprintf(stdout, "\n\n");
	  if( s ) xfree(s);
	  if( t ) xfree(t);
      }
  }

  /* display cards */
  for(i=1; ;i++){
    if( (s = FunParamGets(fun, NULL, i, NULL, &got)) ){
      if( dolist ) {
	if( strncmp(s, "        ", 8) ) {
	  /* John did this -- its not my fault */
	  fprintf(stdout, "%8.8s\t%s\n", s,
		  (char *)ft_cardget((FITSCard)s));
	}
      } else {
	if( dotype )
	  fprintf(stdout, "%c ", got);
	fprintf(stdout, fmt, s);
      }
      xfree(s);
    }
    else{
      break;
    }
  }
  fflush(stdout);
  
  /* close file */
  FunClose(fun);

  /* if we are processing all extensions, go back for more */
  if( doall ){
    snprintf(cname, clen, "%s[%d]", iname, extno);
    extno++;
    goto again;
  }

done:
  /* clean up */
  if( iname ) xfree(iname);
  if( oname ) xfree(oname);
  if( cname ) xfree(cname);
  if( ename ) xfree(ename);

  return(rcode);
}
